### PR TITLE
CORTX-27752:  Rotation of core dump  files.

### PR DESCRIPTION
**Pull request by papan-singh**
Signed-off-by: Papan Kumar Singh <papan.kumarsingh@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by papan-singh** [comment]
this patch will  remove core files sorted by time modified and will preserve 2 core files.
Backup of core file with same PID will be handled in separate ticket .

**Comment by papan-singh** [comment]
patch is tested on k8s
`[root@ssc-vm-rhev4-2206 ~]# kubectl exec -it cortx-data-ssc-vm-rhev4-2204-6cbfd4df7c-56vgm -c cortx-motr-io-001 -- /bin/bash
[root@cortx-data-headless-svc-ssc-vm-rhev4-2204 /]# ls /etc/cortx/motr/m0d-0x7200000000000001\:0x63
core.1647924315.2134  core.1647924379.2296  db  db-log  stobs
[root@cortx-data-headless-svc-ssc-vm-rhev4-2204 /]# ps -ef
UID         PID   PPID  C STIME TTY          TIME CMD
root          1      0  0 Mar21 ?        00:00:00 /bin/sh -c /opt/seagate/cortx/motr/bin/motr_setup start --services ioser
root          7      1  0 Mar21 ?        00:00:32 python3 /opt/seagate/cortx/motr/bin/motr_setup start --services ioservic
root        149      7  0 Mar21 ?        00:00:00 /bin/bash /usr/libexec/cortx-motr/motr-monitor 0x7200000000000001:0x63
root       1011      0  0 Mar21 pts/0    00:00:00 /bin/bash
root       2387    149  8 04:46 ?        00:02:47 /usr/bin/m0d -e libfab:inet:tcp:cortx-data-headless-svc-ssc-vm-rhev4-220
root       2473      0  0 05:17 pts/1    00:00:00 /bin/bash
root       2494   2473  0 05:18 pts/1    00:00:00 ps -ef
[root@cortx-data-headless-svc-ssc-vm-rhev4-2204 /]# kill -SIGSEGV 2387
[root@cortx-data-headless-svc-ssc-vm-rhev4-2204 /]# ls /etc/cortx/motr/m0d-0x7200000000000001\:0x63
core.1647924379.2296  core.1647926314.2387  db  db-log  stobs
`

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            usage
+            ;;
+    esac
+done
+
+if [[ $platform = "virtual" ]]; then
+
+    core_files_max_count=2
+else
+    core_files_max_count=$((core_files_max_count + 2))
```
Comment (nkommuri) [inline scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:90]: Why are we adding +2 for hardware machines?

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                           core directory($core_dir) ##"               
+
+            # get the files sorted by time modified 
+            # (most recently modified comes last), that 
+            # is older files comes first
+            
+            if [[ $platform = "physical" ]]; then
+                peserve_files=$((core_files_max_count - 2))
+                files_to_remove=$(ls -tr "$core_dir" | grep core | \
+                                     head -n -"$peserve_files" | awk 'NR>2')
```
Comment (nkommuri) [inline scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:133]: Why are we considering line numbers greater than 2 only (awk 'NR>2')?

File: scripts/install/usr/libexec/cortx-motr/motr-monitor [hunk 3]
```diff
@@ -44,5 +43,6 @@ while true; do
 		echo "restarting $1"
 		/usr/libexec/cortx-motr/motr-server "m0d-$1" &
 		PID=$!
+                /opt/seagate/cortx/motr/libexec/core_logrotate.sh "m0d-$1" "$PID" &
```
Comment (nkommuri) [inline scripts/install/usr/libexec/cortx-motr/motr-monitor:46]: Do we need to send "m0d-$1" and "$PID" as arguments? core_logrorate.sh script is not accepting these.

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh (hunk 1 above)
Comment (papan-singh) [reply scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:90]: on R2 we can have more core files similar to m0trace

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh (hunk 2 above)
Comment (papan-singh) [reply scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:133]: have  to preserve only latest 2 core file

File: scripts/install/usr/libexec/cortx-motr/motr-monitor (hunk 3 above)
Comment (papan-singh) [reply scripts/install/usr/libexec/cortx-motr/motr-monitor:46]: we can use it to optimize in later

Comment (nkommuri) [reply scripts/install/usr/libexec/cortx-motr/motr-monitor:46]: No need to pass PID to core_logrotate.sh, as we will be handling duplicate PIDs using core filename format. No need to pass unnecessary/unused args.

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh (hunk 2 above)
Comment (nkommuri) [reply scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:133]: What is default number of core files that we need to preserve on VM and H/W.  on VM it is 2, and on H/W it is 5 right?

File: scripts/install/usr/libexec/cortx-motr/motr-monitor (hunk 3 above)
Comment (papan-singh) [reply scripts/install/usr/libexec/cortx-motr/motr-monitor]: incorporated

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh (hunk 2 above)
Comment (papan-singh) [reply scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh]: incorporated

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh (hunk 1 above)
Comment (papan-singh) [reply scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh]: incorporated

**Review by nkommuri — APPROVED** [review]

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+echo "Core file directory: $motr_coredirs"
+
+# check for core directory entries
+for motr_coredir in ${motr_coredirs[@]}; do
+    [[ -d $motr_coredir ]] || {
+        echo "$motr_coredir: no such directory"
+        continue
+    }
+
+    [[ $(check_param "$motr_coredir") = "continue" ]] && continue || echo "$motr_coredir"
```
Comment (yatin-mahajan) [inline scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:101]: does this check_param() function is necessary here ? can we check file/directory here itself ?

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+# script is used to delete the old core files
+# argument1: <number of latest core files to retain>
+# Default number of latest core files is 2
+# ./core_logrotate.sh 5
```
Comment (yatin-mahajan) [inline scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:24]: s/core_logrotate.sh 5/./core_logrotate.sh -n 5/

Comment (papan-singh) [reply scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:24]: incorporated

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            fi
+            ;;
+        *)
+            usage
+            ;;
+    esac
+done
+
+if [[ $platform = "virtual" ]]; then
+
```
[the remark this answers is not in the document]
Comment (papan-singh) [reply scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:87]: incorporated

**Review by yatin-mahajan — APPROVED** [review]
Looks good to me

**Review by mehjoshi — CHANGES_REQUESTED** [review]

File: scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+source "/opt/seagate/cortx/motr/common/cortx_util_funcs.sh"
+
+usage() { 
+        echo "Usage: bash $(basename "$0") [--help|-h] [-n CoreFileCount]"
+        "Retain recent modified files of given count and remove older core files."
+        "where:"
+        "-n   number of latest core files to retain" 
+              "Physical : Default count of core files is 5"
+              "virtual  : Default count of core files is 2"
+        "--help|-h     display this help and exit" 1>&2;
```
[the remark this answers is not in the document]
Comment (mehjoshi) [reply scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:35]: I mean the usage() function. Did you try to print the usage?

[the remark this answers is not in the document]
Comment (papan-singh) [reply scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh:35]: > Don't you need to add '' at the end of each line above for continuation? Is this code tested?

> [root@cortx-data-headless-svc-ssc-vm-rhev4-2204 ~]# /opt/seagate/cortx/motr/libexec/core_logrotate.sh --help
Server type is virtual
        Usage: bash core_logrotate.sh [--help|-h] [-n CoreFileCount]
        Retain recent modified files of given count and remove older core files.
        where: "
        -n   number of latest core files to retain "
              Physical : Default count of core files is 5
              virtual  : Default count of core files is 2
        --help|-h     display this help and exit

[the remark this answers is not in the document]
Comment (papan-singh) [reply scripts/install/opt/seagate/cortx/motr/libexec/core_logrotate.sh]: > I mean the usage() function. Did you try to print the usage?

[root@cortx-data-headless-svc-ssc-vm-rhev4-2204 /]# /opt/seagate/cortx/motr/libexec/core_logrotate.sh --help
Server type is virtual
        Usage: bash core_logrotate.sh [--help|-h] [-n CoreFileCount]
        Retain recent modified files of given count and remove older core files.
        where:
        -n   number of latest core files to retain
              Physical : Default count of core files is 5
              virtual  : Default count of core files is 2
        --help|-h     display this help and exit

**Review by mehjoshi — APPROVED** [review]

**Pull request merged**